python/cudf/cudf/core/_base_index.py

```diff
@@ -608,19 +608,43 @@ def intersection(self, other, sort=False):
                     (1, 'Blue')],
                 )
         """
+        if not can_convert_to_column(other):
+            raise TypeError("Input must be Index or array-like")
+
         if not isinstance(other, BaseIndex):
-            other = cudf.Index(other, name=self.name)
+            other = cudf.Index(
+                other,
+                name=getattr(other, "name", self.name),
+            )
 
         if sort not in {None, False}:
             raise ValueError(
                 f"The 'sort' keyword only takes the values of "
                 f"None or False; {sort} was passed."
             )
 
-        if self.equals(other):
+        if self.equals(other) or not len(self):
+            dtypes = [self.dtype, other.dtype]
+            common_dtype = cudf.utils.dtypes.find_common_type(dtypes)
```

Comment:
It seems like we are implementing this “common dtype with compatibility mode as str” logic in many places in cudf. Do we need a helper function for this? `common_dtype_compatible`? Not sure.

Reply:
Good idea, Added a helper function for this.